dataactcore/scripts/populate_unique_keys.py

```diff
@@ -0,0 +1,188 @@
+                                func.concat('CONT_IDV_', func.coalesce(dap.piid, '-none-'),
+                                            '_', func.coalesce(dap.agency_id, '-none-')))
+            else:
+                # awards
+                if 'award' in types:
+                    update_keys(dap, 'FPDS', key_type, 'award', years,
+                                func.concat(func.coalesce(dap.agency_id, '-none-'), '_',
+                                            func.coalesce(dap.referenced_idv_agency_iden, '-none-'), '_',
+                                            func.coalesce(dap.piid, '-none-'), '_',
+                                            func.coalesce(dap.award_modification_amendme, '-none-'), '_',
+                                            func.coalesce(dap.parent_award_id, '-none-'), '_',
+                                            func.coalesce(dap.transaction_number, '-none-'), '_',))
+                # IDV
+                if 'IDV' in types:
+                    update_keys(dap, 'FPDS', key_type, 'IDV', years,
+                                func.concat(func.coalesce(dap.agency_id, '-none-'), '_-none-_',
+                                            func.coalesce(dap.piid, '-none-'), '_',
+                                            func.coalesce(dap.award_modification_amendme, '-none-'), '_-none-_-none-'))
+
+        # unpublished FABS
+        if 'unpublishedFABS' in models:
+            dafa = DetachedAwardFinancialAssistance
+
+            if key_type == 'award':
+                # record type 1
+                if 'AGG' in types:
+                    update_keys(dafa, 'unpublished FABS', key_type, 'aggregate', None,
+                                func.concat('ASST_AGG_', func.coalesce(dafa.uri, '-none-'), '_',
+                                            func.coalesce(dafa.awarding_sub_tier_agency_c, '-none-')))
+                # record type not 1
+                if 'NON' in types:
+                    update_keys(dafa, 'unpublished FABS', key_type, 'non-aggregate', None,
+                                func.concat('ASST_NON_', func.coalesce(dafa.fain, '-none-'), '_',
+                                            func.coalesce(dafa.awarding_sub_tier_agency_c, '-none-')))
+            else:
+                # All transaction keys are the same format for FABS
+                update_keys(dafa, 'unpublished FABS', key_type, '', None,
+                            func.concat(func.coalesce(dafa.awarding_sub_tier_agency_c, '-none-'), '_',
+                                        func.coalesce(dafa.fain, '-none-'), '_',
+                                        func.coalesce(dafa.uri, '-none-'), '_',
+                                        func.coalesce(dafa.cfda_number, '-none-'), '_',
+                                        func.coalesce(dafa.award_modification_amendme, '-none-')))
+
+        # published FABS
+        if 'publishedFABS' in models:
+            pafa = PublishedAwardFinancialAssistance
```

Comment:
Stupid question I can't tell with my eyes: are `unpublishedFABS` and `publishedFABS` the same code, just different models? If so, I'd consider refactoring but it's ultimately your call.

Reply:
There are a few differences including model and whether we use years or not so the only way to really refactor would be with a loop and some more tweaking. Probably overall not much cleaner code